File: javascript/2024/day15.js
import fs from 'node:fs';
import { toInt, parseNumbers, Grid } from '../util.js';

const [ first, moves ] = fs.readFileSync(0, 'utf-8').trim().split('\n\n');

let rows = first.split('\n').map(row => row.split(''));

let start = null;
let width = rows[0].length;
let height = rows.length;

const grid1 = new Grid(width, height);
const grid2 = new Grid(width * 2, height);


for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
        let c = rows[y][x];

        if (c === '@') {
            start = [x, y];
        }

        if (c === '#') {
            grid1.set([x, y], '#');
            grid2.set([x * 2, y], "#");
            grid2.set([x * 2 + 1, y], "#");
        }

        if (c === 'O') {
            grid1.set([x, y], 'O');
            grid2.set([x * 2, y], "[");
            grid2.set([x * 2 + 1, y], "]");
        }
    }
}

/* PART 1 */
performMoves(moves, attemptMove1, start, grid1);


let part1 = calculateGPS('O', grid1);

console.log(part1);

/* PART 2 */

let [x, y] = start;

performMoves(moves, attemptMove2, [x * 2, y], grid2);

let part2 = calculateGPS('[', grid2);
console.log(part2);

function performMoves(moves, wasMoved, pos, grid) {
    for (let move of moves.split('')) {
        if (move === '\n') {
            continue;
        }

        let dir = toDir(move);
        let next = add(pos, dir);
        let nextChar = grid.at(next);

        if (nextChar === '.') {
            pos = next;
            continue;
        }

        if (nextChar === '#') {
            continue;
        }

        if (wasMoved(next, dir, grid)) {
            pos = next;
        }
    }
}

function attemptMove1(pos, dir, grid) {
    let next = pos;
    let nextChar = grid.at(next);

    do {
        next = add(next, dir);
        nextChar = grid.at(next);
    } while (nextChar === 'O');

    if (nextChar === '#') {
        return false;
    }

    grid.swap(pos, next);

    return true;
}

function attemptMove2(pos, dir, grid) {
    if (isHorizontal(dir)) {
        return attemptMoveHorizontal(pos, dir, grid);
    } else {
        if (canMoveVertical(pos, dir, grid)) {
            moveVertical(pos, dir, grid);

            return true;
        }

        return false;
    }
}

function attemptMoveHorizontal(pos, dir, grid) {
    let next = pos;
    let nextChar = grid.at(next);
    let backtrack = [pos];

    do {
        next = add(next, dir);
        backtrack.push(next);
        nextChar = grid.at(next);
    } while (nextChar === '[' || nextChar === ']');

    if (nextChar === '#') {
        return false;
    }

    nextChar = '.';

    for (let b of backtrack) {
        let c = grid.at(b);
        grid.set(b, nextChar);
        nextChar = c;
    }

    return true;
}

function canMoveVertical(pos, dir, grid) {
    let c = grid.at(pos);

    if (c === '#') {
        return false;
    }

    if (c === '.') {
        return true;
    }

    let [x, y] = pos;

    if (c === '[') {
        x += 1;
    }

    if (c === ']') {
        x -= 1;
    }

    if (canMoveVertical(add(pos, dir), dir, grid)
        && canMoveVertical(add([x,y], dir), dir, grid)
    ) {
        return true;
    }

    return false;
}

function moveVertical(pos, dir, grid) {
    let c = grid.at(pos);

    if (c === '.') {
        return;
    }

    let next = add(pos, dir);

    let [x, y] = pos;

    if (c === '[') {
        x += 1;
    }

    if (c === ']') {
        x -= 1;
    }

    moveVertical(next, dir, grid);
    moveVertical(add([x, y], dir), dir, grid);

    grid.swap(pos, next);
    grid.swap([x, y], add([x, y], dir));
}

function calculateGPS(c, grid) {
    let sum = 0;
    for (let y = 0; y < grid.height; y++) {
        for (let x = 0; x < grid.width; x++) {
            if (grid.at([x, y]) === c) {
                sum += 100 * y + x;
            }
        }
    }

    return sum;
}

function toDir(char) {
    switch (char) {
        case '^':
            return [0, -1];
        case 'v':
            return [0, 1];
        case '>':
            return [1, 0];
        case '<':
            return [-1, 0];

        default:
            throw new Error('Invalid direction');
    }
}

function add([ax, ay], [bx, by]) {
    return [ax + bx, ay + by];
}

function isHorizontal(dir) {
    return dir[0] !== 0;
}

File: c/2024/day15.c
#include <stdio.h>
#include <stdlib.h>
#include <inttypes.h>
#include "../Point.c"


Point_t step1(Point_t pos, Point_t dir, Grid_t grid)
{
    Point_t next = pos;
    uint8_t nextChar = Grid_atPoint(next, grid);

    do {
        next = Point_add(next, dir);
        nextChar = Grid_atPoint(next, grid);
    } while (nextChar == 'O');

    if (nextChar == '#') {
        return pos;
    }

    pos = Point_add(pos, dir);

    Grid_swap(pos, next, &grid);

    return pos;
}


Point_t stepHorizontal(Point_t pos, Point_t dir, Grid_t grid)
{
    Point_t next = pos;
    uint8_t nextChar = Grid_atPoint(next, grid);

    do {
        next = Point_add(next, dir);
        nextChar = Grid_atPoint(next, grid);
    } while (nextChar == '[' || nextChar == ']');

    if (nextChar == '#') {
        return pos;
    }

    pos = Point_add(pos, dir);
    Point_t p = next;

    while (p.x != pos.x) {
        next = Point(p.x - dir.x, p.y);
        Grid_swap(p, next, &grid);
        p = next;
    }

    return pos;
}

int canMoveVertical(Point_t pos, Point_t dir, Grid_t grid)
{
    uint8_t c = Grid_atPoint(pos, grid);

    if (c == '#') {
        return 0;
    }

    if (c == '.') {
        return 1;
    }

    int x = pos.x;
    int y = pos.y;

    if (c == '[') {
        x += 1;
    }

    if (c == ']') {
        x -= 1;
    }

    return canMoveVertical(Point_add(pos, dir), dir, grid)
        && canMoveVertical(Point(x, y + dir.y), dir, grid);
}

void moveVertical(Point_t pos, Point_t dir, Grid_t grid)
{
    uint8_t c = Grid_atPoint(pos, grid);

    if (c == '.') {
        return;
    }

    int x = pos.x;

    if (c == '[') {
        x += 1;
    }

    if (c == ']') {
        x -= 1;
    }

    Point_t next = Point_add(pos, dir);

    moveVertical(next, dir, grid);
    moveVertical(Point(x, pos.y + dir.y), dir, grid);

    Grid_swap(pos, next, &grid);
    Grid_swap(Point(x, pos.y), Point(x, pos.y + dir.y), &grid);
}

Point_t stepVertical(Point_t pos, Point_t dir, Grid_t grid)
{
    Point_t next = Point_add(pos, dir);
    if (canMoveVertical(next, dir, grid)) {
        moveVertical(next, dir, grid);

        return next;
    }

    return pos;
}

Point_t step2(Point_t pos, Point_t dir, Grid_t grid)
{
    if (dir.y == 0) {
        return stepHorizontal(pos, dir, grid);
    } else {
        return stepVertical(pos, dir, grid);
    }

    return pos;
}


int calculateGPS(Grid_t grid, uint8_t c)
{
    int sum = 0;

    for (int y = 0; y < grid.height; y++) {
        for (int x = 0; x < grid.width; x++) {
            if (Grid_at(x, y, grid) == c) {
                sum += y * 100 + x;
            }
        }
    }

    return sum;
}

int main()
{
    int row = 0;
    int col = 0;
    int width = 0;
    int height = 0;

    // Parse first row to find the size of the grid
    while (!feof(stdin)) {
        int c = fgetc(stdin);

        if (c == '\n') {
            // Assume the input grid is square
            width = col;
            height = col;
            row += 1;
            col = 0;
            break;
        }

        col += 1;
    }

    Grid_t grid1 = Grid(width, height);
    Grid_t grid2 = Grid(width * 2, height);

    for (int x = 0; x < width; x++) {
        Grid_set(Point(x, 0), '#', &grid1);
        Grid_set(Point(x * 2, 0), '#', &grid2);
        Grid_set(Point(x * 2 + 1, 0), '#', &grid2);
    }

    // Parse the rest of the grid part of the input
    Point_t start = Point(0, 0);

    while (!feof(stdin)) {
        int c = fgetc(stdin);

        if (c == '\n' && col == 0) {
            // Reached the end of the grid part of the input.
            break;
        }

        if (c == '\n') {
            row += 1;
            col = 0;
            continue;
        }

        if (c == '@') {
            start = Point(col, row);
            c = '.';
        }

        Grid_set(Point(col, row), c, &grid1);

        if (c == 'O') {
            Grid_set(Point(col * 2, row), '[', &grid2);
            Grid_set(Point(col * 2 + 1, row), ']', &grid2);
        } else {
            Grid_set(Point(col * 2, row), c, &grid2);
            Grid_set(Point(col * 2 + 1, row), c, &grid2);
        }


        col += 1;
    }


    // Perform all the moves

    Point_t pos1 = start;
    Point_t pos2 = Point(start.x * 2, start.y);

    while (!feof(stdin)) {
        int c = fgetc(stdin);

        if (c == EOF || c == '\n') {
            continue;
        }

        Direction_t dir = Direction_fromArrow(c);
        Point_t delta = Direction_toPoint(dir);

        pos1 = step1(pos1, delta, grid1);
        pos2 = step2(pos2, delta, grid2);
    }


    int part1 = calculateGPS(grid1, 'O');
    int part2 = calculateGPS(grid2, '[');

    printf("%d\n", part1);
    printf("%d\n", part2);

    Grid_free(grid1);
    Grid_free(grid2);
}


